### PR TITLE
Add ToolBox extension

### PR DESCRIPTION
**Pull request by kkange**
## Description

This is a collection extension of various string conversion scripts.
You can use the strings stored in the clipboard to execute the function or you can directly enter it and execute it.

Scripts are easy to add and we will continue to add more.

## Type of change

- New extension

## Screencast

[script list](https://imgur.com/XidJIIQ)
[form input](https://imgur.com/Vzqy0J4)
[direct input](https://imgur.com/pUfNOaY)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I checked that extension's icon has size of `512x512`

### CODE REVIEW

File: extensions/toolbox/src/script/util.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { execa } from "execa";
+
+export async function readClipboard() {
+  const { stdout } = await execa("pbpaste", ["unicorns"]);
```
Comment (es183923) [inline extensions/toolbox/src/script/util.ts:4]: why is there a `["unicorns"]` here?

**Comment by es183923** [comment]
I just want to say that some of the functions are already available in other extensions, such as the Came Case function, in the [change case](https://www.raycast.com/es183923/change-case) extension, and the random number function, in the [random data generator](https://www.raycast.com/loris/random) extension.

File: extensions/toolbox/src/script/util.ts (hunk 1 above)
Comment (kkange) [reply extensions/toolbox/src/script/util.ts]: Thanks to your advice, I modified it.

**Comment by kkange** [comment]
> I just want to say that some of the functions are already available in other extensions, such as the Camel Case function, in the [change case](https://www.raycast.com/es183923/change-case) extension, and the random number function, in the [random data generator](https://www.raycast.com/loris/random) extension.


The toolbox extension program is a simple text conversion tool with various functions but few choices to set.

On the other hand, the extended programs you told me have a variety of options and various data, so I think they are different from the toolbox expansion.

File: extensions/toolbox/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+              title={"Run Script Input"}
+              icon={Icon.Text}
+              onAction={async () => {
+                action({ useClipboard: false, isWindowClose: false });
+              }}
+            />
+          )}
+          {(info.type === "all" || info.type === "clipboard") && (
+            <ActionPanel.Item
+              shortcut={{ modifiers: ["cmd"], key: "s" }}
```
Comment (es183923) [inline extensions/toolbox/src/index.tsx:115]: I think cmd+shift-enter would be better here.

File: extensions/toolbox/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+              title={"Run Script - Window Keep"}
```
Comment (es183923) [inline extensions/toolbox/src/index.tsx:116]: A better title would be `Run Script - Keep Window`

File: extensions/toolbox/src/script/date/timestamp.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { Script } from "../type";
+
+export const toTimestamp: Script = {
+  info: {
+    title: "Date to Timestamp",
+    desc: "Converts dates to Unix timestamp",
```
Comment (es183923) [inline extensions/toolbox/src/script/date/timestamp.ts:6]: `"Convert dates to Unix timestamps"` would be better here.

File: extensions/toolbox/src/script/date/utc.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Script } from "../type";
+
+export const toUtc: Script = {
+  info: {
+    title: "Date to UTC",
+    desc: "Converts dates and timestamps to UTC dates",
```
Comment (es183923) [inline extensions/toolbox/src/script/date/utc.ts:6]: `Convert dates and timestamps to UTC dates` would be better here.

File: extensions/toolbox/src/script/string/camelCase.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { Script } from "../type";
+import * as _ from "lodash";
+
+export const camelCase: Script = {
+  info: { title: "Camel Case", desc: "Converts Your Text To Camel Case", type: "all", example: "ray cast" },
```
Comment (es183923) [inline extensions/toolbox/src/script/string/camelCase.ts:5]: `Converts Your Text To Camel Case` should be `Convert your text into Camel Case`.

**Comment by es183923** [comment]
Instead of saying `Decodes` or `Converts`, it should just be `Decode` or `Convert`

File: extensions/toolbox/src/index.tsx (hunk 3 above)
Comment (kkange) [reply extensions/toolbox/src/index.tsx:116]: Thank you. I fixed it.

File: extensions/toolbox/src/script/date/timestamp.ts (hunk 4 above)
Comment (kkange) [reply extensions/toolbox/src/script/date/timestamp.ts:6]: Thank you. I fixed it.

File: extensions/toolbox/src/script/date/utc.ts (hunk 5 above)
Comment (kkange) [reply extensions/toolbox/src/script/date/utc.ts:6]: Thank you. I fixed it.

File: extensions/toolbox/src/script/string/camelCase.ts (hunk 6 above)
Comment (kkange) [reply extensions/toolbox/src/script/string/camelCase.ts:5]: Thank you. I fixed it.

**Comment by kkange** [comment]
> Instead of saying `Decodes` or `Converts`, it should just be `Decode` or `Convert`

Thank you for helping me. You're a good person.

File: extensions/toolbox/src/index.tsx (hunk 2 above)
Comment (kkange) [reply extensions/toolbox/src/index.tsx:115]: Oh, that's right. If you change it like that, it'll be consistent.

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @kkange 👋 

I tested your extension and have a few comments:

- If having nothing or wrong format in clipboard while visiting “Date to Timestamp” and “Date to UTC” it flashes loading toast → error toast and change view.

- The UI is a bit confusing when nothing is explaining to me what to do, I didn't catch what I should do at first — the empty item could hold some help (I can also see that every view have a form view, I would stick with that one for the most of them instead)

- The loading toast when entering a script feels redundant to me, since it's not loading anything that should cause delay and only flashes quickly.

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by kkange** [comment]
Thank you for the review.

I think I revised what you said.

File: extensions/toolbox/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    <List.Item
+      title={info.title}
+      accessoryTitle={info.desc}
+      keywords={keyword}
+      icon={info.icon}
+      actions={
+        <ActionPanel>
+          {(info.type === "all" || info.type === "clipboard") && (
+            <ActionPanel.Item
+              title={"Run Script Clipboard"}
```
Comment (pernielsentikaer) [inline extensions/toolbox/src/index.tsx:129]:
```suggestion
              title={"Run Script to Clipboard"}
```

File: extensions/toolbox/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+              icon={Icon.ArrowRight}
+              onAction={async () => {
+                action("clipboard");
+              }}
+            />
+          )}
+          {(info.type === "all" || info.type === "noclipboard" || info.type === "direct") && (
+            <ActionPanel.Item
+              title={"Run Script Direct"}
```
Comment (pernielsentikaer) [inline extensions/toolbox/src/index.tsx:138]:
```suggestion
              title={"Run Script"}
```

File: extensions/toolbox/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    !content.isError &&
+    content.result.length > 0 && (
+      <ActionPanel title={info.title}>
+        <ActionPanel.Section>
+          <ActionPanel.Item
+            title="Copy Result to Clipboard"
+            icon={Icon.Clipboard}
+            onAction={async () => {
+              await copyTextToClipboard(content.result);
+              await showHUD("Copy Result");
```
Comment (pernielsentikaer) [inline extensions/toolbox/src/index.tsx:226]:
```suggestion
              await showHUD("✅ Result Copied to Clipboard");
```

File: extensions/toolbox/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+            }}
+          />
+
+          <ActionPanel.Item
+            title="Copy Query to Clipboard"
+            icon={Icon.Clipboard}
+            onAction={async () => {
+              await copyTextToClipboard(content.query);
+              await showHUD("Copy Query");
```
Comment (pernielsentikaer) [inline extensions/toolbox/src/index.tsx:235]:
```suggestion
              await showHUD("✅ Query Copied to Clipboard");
```

File: extensions/toolbox/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        scriptResult = await runScript(query);
+        if (scriptResult.isSuccess) {
+          await copyTextToClipboard(scriptResult.result);
+        }
+      }
+      _.delay(async () => {
+        isClipboardScriptRunning = false;
+        toast.hide();
+        if (scriptResult.isSuccess) {
+          await showHUD("Copy Result");
```
Comment (pernielsentikaer) [inline extensions/toolbox/src/index.tsx:106]:
```suggestion
          await showHUD("✅ Result Copied to Clipboard");
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @kkange 👋 

Firstly, sorry for the longer response time, I had to try it out for a longer period than expected.
 
Thanks for the update, we're getting closer to release, I have some text corrections and comments:

- Adding extra delay makes the extension “slow” — since everything is happening locally there is no really need for loading spinners, I suggest removing extra delay and loading spinners.

Looking forward for the update 🔥 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by kkange** [comment]
I accepted your proposal and changed it.

And after using it for a few days,
It was convenient to change the basic action to the input screen. So, I changed it.

Action order.
1. Run Script
2. Run Script From
3. Run Script to Clipboard

**Review by pernielsentikaer — APPROVED** [review]
Hi @kkange 👋 

I'm going to accept it now, I think its good as it is right now 💯 

Thanks for you contribution 👍

**Pull request merged**